Tag Processor: throw when supplied unacceptible attribute names.

dmsnell opened the pull request:
Part of #44410 

## What

The `WP_HTML_Tag_Processor` allows setting new HTML attributes with a given name and value. Previously this has allowed any string input for the attribute name, but we have to be careful not to print output that might break the HTML we're modifying.

In this patch we're adding a check against the given attribute name and rejecting invalid or unacceptible names. WordPress here is more restrictive than HTML5.

## Why?

We don't want to allow setting attributes whose names could break the HTML syntax or open new vulnerabilities in code that parses post content.

## How?

Validate the given attribute name with a PCRE pattern that looks for unalloyed characters.

## Testing Instructions

Please audit the code and review the new unit tests.
Review with an eye towards security.

cc: @azaozz

dmsnell committed:
Tag Processor: throw when supplied unacceptible attribute names.

The `WP_HTML_Tag_Processor` allows setting new HTML attributes with
a given name and value. Previously this has allowed any string input
for the attribute name, but we have to be careful not to print output
that might break the HTML we're modifying.

In this patch we're adding a check against the given attribute name
and rejecting invalid or unacceptible names. WordPress here is more
restrictive than HTML5.

adamziel commented:
~~Would you please add a docstring link to the docs section where these characters are specified?~~ It was there, I just missed it!

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -948,6 +948,38 @@ public function set_attribute( $name, $value ) {
+		 * @see https://html.spec.whatwg.org/#attributes-2
+		 */
+		if ( preg_match(
+			'~[ "\'>&</=\x{00}-\x{1F}\x{FDD0}-\x{FDEF}\x{FFFE}\x{FFFF}\x{1FFFE}\x{1FFFF}\x{2FFFE}\x{2FFFF}\x{3FFFE}\x{3FFFF}\x{4FFFE}\x{4FFFF}\x{5FFFE}\x{5FFFF}\x{6FFFE}\x{6FFFF}\x{7FFFE}\x{7FFFF}\x{8FFFE}\x{8FFFF}\x{9FFFE}\x{9FFFF}\x{AFFFE}\x{AFFFF}\x{BFFFE}\x{BFFFF}\x{CFFFE}\x{CFFFF}\x{DFFFE}\x{DFFFF}\x{EFFFE}\x{EFFFF}\x{FFFFE}\x{FFFFF}\x{10FFFE}\x{10FFFF}]~Ssu',
```
Is this any different than `strcspn`? It seems to support unicode code points:

```
php > var_dump(strcspn("ab\u{CFFFE}c", "\u{CFFFE}"));
int(2)
```

adamziel replied:
Ah I see, you used a few ranges here – nice! Here's my understanding of the regexp:

* ` "\'>&</=` are specifically forbidden by https://html.spec.whatwg.org/#attributes-2
* Ditto for controls: `\x{00}-\x{1F}`
* Ditto for noncharacters: `\x{FDD0}-\x{FDEF}`
* The rest of the expression is explicitly listed in the [noncharacters section](https://infra.spec.whatwg.org/#control). 

This looks good to me!

dmsnell replied:
to be honest I assumed it wouldn't work with `strcspn` and didn't even try the Unicode code points. just now I confirmed on PHP 5.4.40 and ~it worked as expected - I'll update~.

originally I had the control codes listed with `strcspn` but then moved to the `preg_match` for the ranges.

**update**: this _doesn't_ work as expected because it breaks apart the Unicode characters into their byte sequences, thus rejecting any string containing any of the bytes in that sequence

dmsnell replied:
@adamziel we might need to retain the `preg_match` because I think your test code is subtly wrong. as I suspected `strspn` and `strcspn` operate on normal PHP strings, which are byte-wise compared. what you're seeing is a rejection of strings containing any of the bytes `\xF3\x8F\xBF\xBE`, not those containing only that sequence.

```php
php > var_dump( strcspn( "test\u{CFFFE}test", "\u{CFFFE}" ) )
int(4)

php > var_dump( strcspn( "test\u{CFFFF}test", "\u{CFFFE}" ) )
int(4)

php > var_dump( strcspn( "test\x8ftest", "\u{CFFFE}" ) );
int(4)

php > var_dump( strcspn( "testtest", "\u{CFFFE}" ) );
int(8)
```

adamziel replied:
Oh this is such a good spot 💯  Let's stick to `preg_match`, then.

adamziel commented:
I needed to manually set `WP_DEBUG` to true to make the tests pass as the exception is only thrown conditionally.

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -948,6 +948,38 @@ public function set_attribute( $name, $value ) {
+				throw new Exception( 'Invalid attribute name' );
+			}
+
+			return;
```
return new WP_Error perhaps?

gziolo replied:
I don't think we should return an object as devs won't check the return value anyway.

The special case in the debug mode looks like a reasonable approach. I'm curious whether there is any standardized way of logging those issues. 

Maybe we could call [`_doing_it_wrong`](https://developer.wordpress.org/reference/functions/_doing_it_wrong/). We use that in several places when registering blocks or similar stuff for the block editor, for example:

https://github.com/WordPress/wordpress-develop/blob/2b1febd20d77898eb81439a688ea5597da00172a/src/wp-includes/class-wp-block-type-registry.php#L55L62

dmsnell replied:
this is the big point about this patch that I'm unsure of. my only goal is to quickly bail if someone submits the wrong attribute, as I don't think it's likely people will send user-input attribute names here. if they _do_ do that then I don't want it to crash the full render.

I'll ask around and see if anyone else has experience with this kind of failure.

azaozz replied:
> my only goal is to quickly bail if someone submits the wrong attribute

Yes, think this is the best approach here. If a plugin allows user input for attribute names (very unlikely as usually these are not random), it will have to make sure the input is valid, or the attribute will be skipped.

azaozz replied:
> I'm curious whether there is any standardized way of logging those issues.

Think there isn't one, unfortunately. Has been discussed few times afaik but a decision on how to implement it was not reached. Thinking it is time for this to be added to core. Perhaps another constant: `WP_DEV_MODE` or similar, then be much bolder about throwing errors and exceptions (with backtrace?) and writing in logs.

Using `WP_DEBUG` seems proper here imho. `doing_it_wrong` doesn't seem as good because it is targeted more at developers that try to use some function/method improperly (when there are better ways).

gziolo replied:
@azaozz, great feedback! Thank you so much for clarifying where given options fit best.

adamziel commented on phpunit/html/WP_HTML_Tag_Processor_Test.php:
```diff
@@ -248,6 +248,57 @@ public function test_set_attribute_with_a_non_existing_attribute_adds_a_new_attr
+			'noncharacters_4'  => array( html_entity_decode( 'te&#xDFFFF;st' ) ),
+			'noncharacters_5'  => array( html_entity_decode( '&#x10FFFE;' ) ),
+			'wp_no_lt'         => array( 'id<script'),
+			'wp_no_amp'        => array( 'class&lt;script'),
```
Adding one data entry per each noncharacter and each control character would prevent any and all future regressions should we update the implementation to `strcspn` or anything else. This is a great start, but leaves a lot of ground uncovered.

dmsnell replied:
it does but also I was trying to balance exhaustiveness of the test suite against practicality of it. this one grabs a handful of characters in the ranges and smoke-tests them. if we wanted to test everything (which seems like a goal with diminishing returns) then we could run loops in the test to generate the rejected input.

adamziel replied:
Alright, that's fair – let's stick to this. I just added some positive test examples to ensure that correct attribute names are accepted.

adamziel reviewed (approved):
LGTM! I left a few comments, but my only substantial point is about the test cases.

dmsnell committed:
Update code comments

dmsnell committed:
Split string into multi-line, move tests into isolated runner

The tests verifying behaviors dependent on whether `WP_DEBUG` is
set need to run in an isolated environment so that the constant's
definition doesn't bleed into other tests. The isolated runner is
very slow in comparison to the non-isolated runner so only those
tests dependent on the constant are moved over.

adamziel committed:
Test for valid attributes

adamziel committed:
Lint

adamziel commented:
I added some positive test cases in ff9da888d36a98d686b612d7a8926b1fdbe7259b.

adamziel commented on phpunit/html/WP_HTML_Tag_Processor_Isolated_Test.php:
```diff
@@ -0,0 +1,145 @@
+	}
+
+	/**
+	 * Data provider with invalid HTML attribute names.
+	 *
+	 * @return array {
+	 *     @type string $attribute_name Text considered invalid for HTML attribute names.
+	 * }
```
```suggestion
	 * Data provider with valid HTML attribute names.
	 *
	 * @return array {
	 *     @type string $attribute_name Text considered valid for HTML attribute names.
	 * }
```

adamziel committed:
Update phpunit/html/WP_HTML_Tag_Processor_Isolated_Test.php

adamziel committed:
Lint
